pkg/action/utils.go

```diff
@@ -0,0 +1,94 @@
+/*
+Copyright SUSE LLC.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package action
+
+import (
+	"github.com/pkg/errors"
+	"helm.sh/helm/v3/pkg/action"
+	"helm.sh/helm/v3/pkg/chart"
+	"strings"
+)
+
+// SetNamespace sets the Namespace that should be used based on annotations or fallback to default
+// both on the action and on the storage.
+// if setDefault is true then it will just set the default namespace
+// This will read the chart annotations. If no annotations, it leave the existing ns in the action.
+// targetNS can be either the default namesapce (usually "default") or the namespace passed via
+// cli flag
+func SetNamespace(x interface{}, chart *chart.Chart, targetNS string, NamespaceFromFlag bool) {
+	namespace := targetNS
+	// NamespaceFromFlag is mainly used when we set the namespace via the cli flag --namespace
+	// and it has priority over everything else
+	if NamespaceFromFlag {
+		namespace = targetNS
+	} else {
+		if chart.Metadata.Annotations != nil {
+			if val, ok := chart.Metadata.Annotations["hypper.cattle.io/namespace"]; ok {
+				namespace = val
+			} else {
+				if val, ok := chart.Metadata.Annotations["catalog.cattle.io/namespace"]; ok {
+					namespace = val
+				}
+			}
+		}
+	}
+
+	switch i := x.(type) {
+	case *Install:
+		i.Namespace = namespace
+		i.Config.SetNamespace(namespace)
```

Comment:
Here's an idea. This bit of functionality seems to repeat. Just this small part.

What about moving it to a `SetNamespace` property on `Upgrade` and `Install`. Then, create an interface with one method of `SetNamespace` that takes the namespace and does just this work. In the arguments for this function you can use that interface instead of `interface{}`. Then you don't have to do a type switch and can just call `i.SetNamespace(namespace)` here.

The compiler would make sure everything has the property.

You can also use composition to make that `SetNamespace` function once and embed it in the `Install` and `Upgrade` structs. Or, you could embed all of this that way.

Reply:
Wouldn't that mean that:
 1 - We are still repeating code
 2 - We have 2 places to change the code if we find a bug

?

I like the second suggestion thought, seems more nice and easy to do with minimal changes!

Reply:
Tried both suggestions, Im incapable of making it work as expected without messing it up or making it twice as complex...


Any suggestions or code examples? @mattfarina @viccuad @dragonchaser